core/src/builder/core.rs

```diff
@@ -166,12 +167,12 @@ where
 		at: Option<BlockId<Block>>,
 		exec: impl FnOnce() -> R,
 	) -> Result<R, String> {
-		let (state, at) = if let Some(req_at) = at {
-			(self.backend.state_at(req_at), req_at)
-		} else {
-			let at = BlockId::Hash(self.client.info().best_hash);
-			(self.backend.state_at(at.clone()), at)
+		let hash = match at {
+			Some(BlockId::Hash(req_at)) => req_at,
+			_ => self.client.info().best_hash,
 		};
+		let state = self.backend.state_at(&hash);
+		let at = sp_api::BlockId::Hash(hash);
```

Comment:
This is wrong. If a user uses `BlockId::Number(...)` then this match call will use `self.client.into().best_hash` instead. 

If you find a prettier way to do this, I am happy to see that in 😄

Reply:
I've added a case for BlockId::Number as well :).  There was an update to Substrate witch necessitated hash grabbing for this -- I think it was this if I recall correctly https://github.com/paritytech/substrate/commit/f3139874cb50f9028ecba9bdbd3004e7f3f228f5.